src/platforms/wayland/wl_egl_display_provider.cpp

```diff
@@ -12,47 +12,132 @@ namespace mg = mir::graphics;
 namespace mgw = mir::graphics::wayland;
 namespace geom = mir::geometry;
 
+namespace
+{
+// Utility to restore EGL state on scope exit
+class CacheEglState
+{
+public:
+    CacheEglState() = default;
+
+    ~CacheEglState()
+    {
+        eglMakeCurrent(dpy, draw_surf, read_surf, ctx);
+    }
+private:
+    CacheEglState(CacheEglState const&) = delete;
+    EGLDisplay const dpy = eglGetCurrentDisplay();
+    EGLContext const ctx = eglGetCurrentContext();
+    EGLSurface const draw_surf = eglGetCurrentSurface(EGL_DRAW);
+    EGLSurface const read_surf = eglGetCurrentSurface(EGL_READ);
+};
+}
+
+class mgw::WlDisplayAllocator::SurfaceState
+{
+public:
+    SurfaceState(EGLDisplay dpy, struct ::wl_egl_window* wl_window) :
+        dpy{dpy}, wl_window{wl_window} {}
+
+    ~SurfaceState()
+    {
+        if (egl_surface != EGL_NO_SURFACE) eglDestroySurface(dpy, egl_surface);
+        wl_egl_window_destroy(wl_window);
+    }
+
+    auto surface(EGLConfig egl_config) -> EGLSurface
+    {
+        std::lock_guard lock{mutex};
+        if (egl_surface == EGL_NO_SURFACE)
+        {
+            egl_surface = eglCreatePlatformWindowSurface(dpy, egl_config, wl_window, nullptr);
+        }
+
+        if (egl_surface == EGL_NO_SURFACE)
+        {
+            BOOST_THROW_EXCEPTION(egl_error("Failed to create EGL surface"));
+        }
+
+        return egl_surface;
+    }
+
+private:
+    EGLDisplay const dpy;
+    struct ::wl_egl_window* const wl_window;
+
+    std::mutex mutex;
+    EGLSurface egl_surface{EGL_NO_SURFACE};
+};
+
 class mgw::WlDisplayAllocator::Framebuffer::EGLState
 {
 public:
-    EGLState(EGLDisplay dpy, EGLContext ctx, EGLSurface surf)
+    EGLState(EGLDisplay dpy, EGLContext ctx, EGLSurface surf, std::shared_ptr<SurfaceState> ss)
         : dpy{dpy},
           ctx{ctx},
-          surf{surf}
+          surf{surf},
+          ss{std::move(ss)}
     {
+        CacheEglState stash;
+
+        make_current();
+        // Don't block in eglSwapBuffers; we rely on external synchronisation to throttle rendering
+        eglSwapInterval(dpy, 0);
+        release_current();
     }
 
     ~EGLState()
     {
+        CacheEglState stash;
```

Comment:
I hadn't realised that could matter. But in testing this suggestion I found that there's still an intermittent failure mode (with or without this):

```
terminate called after throwing an instance of 'boost::wrapexcept<std::system_error>'
  what():  Failed to create EGL surface: EGL_BAD_DISPLAY (0x3008)
```

Reverted to "Draft" as I need to track that down.

Reply:
Oh! This may be worse than I thought...
```
Fatal glibc error: pthread_mutex_lock.c:94 (___pthread_mutex_lock): assertion failed: mutex->__data.__owner == 0
```

Reply:
At least I now know how to trigger this failure - it's when the host reaches "idle-timeout"

Reply:
@RAOF I have updated `~EGLState()` to something I think is simpler and should work. If you could confirm that?

Meantime, there's another failure mode I have still to resolve

Reply:
> I hadn't realised that could matter.

*Mostly* it doesn't matter - the call will fail, but without any side-effect. *Except* when someone tries with EGL debug enabled, at which point the failure will be logged. My only concern here was avoiding a potential debugging wild goose chase in the future.

'Twas only a happy accident that this prompted further testing that identified your new problem!

> I have updated ~EGLState() to something I think is simpler and should work. If you could confirm that?

Yup! That works nicely.